cpp/src/sort/sort.cu

```diff
@@ -19,13 +19,71 @@
 #include <cudf/detail/sorting.hpp>
 #include <cudf/sorting.hpp>
 #include <cudf/table/table_view.hpp>
+#include <cudf/unary.hpp>
 
 #include <sort/sort_impl.cuh>
 
 #include <rmm/cuda_stream_view.hpp>
 
 namespace cudf {
 namespace detail {
+
+// Convert null_mask to BOOL8 columns and flatten the struct children in order.
+void flatten_struct_column(column_view const& col,
+                           std::vector<column_view>& all_columns,
+                           std::vector<std::unique_ptr<column>>& validity_as_column)
+{
+  for (auto it = col.child_begin(); it != col.child_end(); it++) {
+    auto const& child = *it;
+    if (child.type().id() == type_id::STRUCT) {
+      if (child.nullable()) {
```

Comment:
Here you only consider the case when a child column is struct and nullable. Does this still work if the child is not a struct but it is nullable? How about a string column?

Reply:
A child that is a nullable struct is the only case that matters here as the null mask for the validity of the structs is independent of the null masks for the children and so that is the only thing that needs to be expanded.